src/Npgsql/Internal/TypeHandlers/NumericHandlers/NumericHandler.cs

```diff
@@ -157,77 +264,172 @@ public override int ValidateAndGetLength(decimal value, NpgsqlParameter? paramet
+            => Write((decimal)value, buf, lengthCache, parameter, async, cancellationToken);
         /// <inheritdoc />
-        public void Write(float value, NpgsqlWriteBuffer buf, NpgsqlParameter? parameter)  => Write((decimal)value, buf, parameter);
+        public Task Write(float value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async, CancellationToken cancellationToken = default)
+            => Write((decimal)value, buf, lengthCache, parameter, async, cancellationToken);
         /// <inheritdoc />
-        public void Write(double value, NpgsqlWriteBuffer buf, NpgsqlParameter? parameter) => Write((decimal)value, buf, parameter);
+        public Task Write(double value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async, CancellationToken cancellationToken = default)
+            => Write((decimal)value, buf, lengthCache, parameter, async, cancellationToken);
+
+        static ushort[] FromBigInteger(BigInteger value)
+        {
+            var str = value.ToString(CultureInfo.InvariantCulture);
+            ushort[] result;
+            if (str == "0")
+            {
+                result = new ushort[4];
+            }
+            else
+            {
+                var negative = str[0] == '-';
+                var strLen = str.Length;
+                var numGroups = (strLen - (negative ? 1 : 0) + 3) / 4;
+
+                if (numGroups > 131072 / 4)
+                    throw new InvalidCastException("Cannot write a BigInteger with more than 131072 digits");
+
+                result = new ushort[4 + numGroups];
+
+                var strPos = strLen - numGroups * 4;
+
+                var firstDigit = 0;
+                for (var i = 0; i < 4; i++)
+                {
+                    if (strPos >= 0 && str[strPos] != '-')
+                        firstDigit = firstDigit * 10 + (str[strPos] - '0');
+                    strPos++;
+                }
+
+                result[4] = (ushort)firstDigit;
+
+                for (var i = 1; i < numGroups; i++)
+                {
+                    result[4 + i] = (ushort)((((str[strPos++] - '0') * 10 + (str[strPos++] - '0')) * 10 + (str[strPos++] - '0')) * 10 +
+                                             (str[strPos++] - '0'));
+
+                }
+
+                var lastNonZeroDigitPos = numGroups - 1;
+                while (result[4 + lastNonZeroDigitPos] == 0)
+                    lastNonZeroDigitPos--;
+
+                result[0] = (ushort)(lastNonZeroDigitPos + 1); // number of items in array
+                result[1] = (ushort)(numGroups - 1); // weight
+                result[2] = (ushort)(negative ? SignNegative : SignPositive);
+                result[3] = 0; // dscale
+            }
+
+            return result;
+        }
+
+        public int ValidateAndGetLength(BigInteger value, ref NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter)
```

Comment:
I don't think this is needed here, since the length can be retrieved immediately from `ConvertedValue`, if present. Will this always work or did I miss some case?

Reply:
Yeah, as above, ConvertedValue only works for top-level parameters. But it's not the end of the world if non-top-level are slower.